update 'buildah images' and 'buildah rmi' commands

14rcole opened the pull request:
add more flags to `buildah images` and `buildah rmi`, and write tests

rhatdan commented:
bot, add author to whitelist

rhatdan commented:
bot, test pull request

rhatdan commented:
@14rcole Sign your image.  This is a patch for buildah not kpod, so change the title.

rhatdan commented on cmd/buildah/images.go:
```diff
@@ -48,38 +83,293 @@ func imagesCmd(c *cli.Context) error {
+	if len(c.Args()) == 1 {
+		name = c.Args().Get(0)
+	} else if len(c.Args()) > 1 {
+		return errors.New("'kpod images' requires at most 1 argument")
```
kpod?

rhatdan commented on cmd/buildah/rmi_test.go:
```diff
@@ -0,0 +1,168 @@
+
+func TestProperImageRefTrue(t *testing.T) {
+	// Pull an image so we know we have it
+	cmd := exec.Command("crioctl", "image", "pull", "busybox:latest")
```
crioctl is not part of buildah.

rhatdan commented on cmd/buildah/rmi_test.go:
```diff
@@ -0,0 +1,168 @@
+
+func TestProperImageRefFalse(t *testing.T) {
+	// Pull an image so we know we have it
+	cmd := exec.Command("crioctl", "image", "pull", "busybox:latest")
```
You should probably use the from calls for this or internal calls to storage.

TomSweeneyRedHat commented on cmd/buildah/common_test.go:
```diff
@@ -0,0 +1,88 @@
+func TestGetStore(t *testing.T) {
+	u, err := user.Current()
+	if err != nil {
+		t.Log("Could not determine user.  Running as root may cause tests to fail")
```
Here you're saying you can't run as root, but at line 19 you're saying you must run at root in the else for this if?

14rcole replied:
It should all require root.  I'll pull all of those checks out into another function so that I don't have as much duplicate code running around

TomSweeneyRedHat commented on cmd/buildah/common_test.go:
```diff
@@ -0,0 +1,88 @@
+func TestParseMetadata(t *testing.T) {
+	u, err := user.Current()
+	if err != nil {
+		t.Log("Could not determine user.  Running as root may cause tests to fail")
```
same root/not root bit here.

TomSweeneyRedHat commented on cmd/buildah/rmi.go:
```diff
@@ -43,55 +57,38 @@ func rmiCmd(c *cli.Context) error {
+					}
+					for i := 0; i < len(containers); i++ {
+						if containers[i].ImageID == image.ID {
+							fmt.Printf("Found the container")
```
Not real helpful to have 1 to n "Found the container" 's.  If you think it's important to note, perhaps change this to something like fmt.Printf("Deleting container %s", containers[i].ID)

14rcole replied:
Good catch.  That was actually a debugging output that I had forgotten to delete

rhatdan commented:
@14rcole tests are crashing.

14rcole commented:
My most recent addition to this PR should resolve #133

rhatdan commented:
@14rcole tests crashing.

TomSweeneyRedHat commented on cmd/buildah/common_test.go:
```diff
@@ -0,0 +1,99 @@
+	}
+}
+
+func failTestIfNotRoot(t *testing.T) {
```
This is a good, but I'm sure finger numbing change.  Thanks for making it.

rhatdan commented:
LGTM
@nalind PTAL

TomSweeneyRedHat commented on cmd/buildah/rmi.go:
```diff
@@ -35,75 +47,192 @@ func rmiCmd(c *cli.Context) error {
+	return ctrIDs, nil
+}
+
+func removeRunningContainers(ctrIDs []string, store storage.Store) error {
```
Should this be named 'removeContianers'?  I'm not seeing any checks to see if the container is running within the function.

14rcole replied:
Yes, good catch

TomSweeneyRedHat commented on cmd/buildah/rmi.go:
```diff
@@ -35,75 +47,192 @@ func rmiCmd(c *cli.Context) error {
+		}
+		return nil, fmt.Errorf("error confirming presence of image %q: %v", transports.ImageName(ref), err)
+	}
+	return nil, fmt.Errorf("error parsing %q as a store reference: %v", id, err)
```
should this be "image reference" rather than "store reference"?

TomSweeneyRedHat commented on cmd/buildah/rmi.go:
```diff
@@ -35,75 +47,192 @@ func rmiCmd(c *cli.Context) error {
+		}
+		return nil, fmt.Errorf("error confirming presence of image %q: %v", transports.ImageName(ref), err)
+	}
+	return nil, fmt.Errorf("error parsing %q as a store reference: %v", id, err)
```
I'd suggest "storage image reference" over "store reference"

TomSweeneyRedHat commented on cmd/buildah/rmi.go:
```diff
@@ -35,75 +47,192 @@ func rmiCmd(c *cli.Context) error {
+			img.Close()
+			return ref, nil
+		}
+		return nil, fmt.Errorf("error confirming presence of image %q: %v", transports.ImageName(ref), err)
```
change image to "storrage image reference" perhaps?

TomSweeneyRedHat commented on cmd/buildah/rmi.go:
```diff
@@ -35,75 +47,192 @@ func rmiCmd(c *cli.Context) error {
+			img.Close()
+			return ref, nil
+		}
+		return nil, fmt.Errorf("error confirming presence of image %q: %v", transports.ImageName(ref), err)
```
First error is 'image", second is "image reference".  I'd prefer "storage image reference" for both, but will settle for consistency.

rhatdan commented:
@14rcole Please rebase.

14rcole commented:
Rebasing didn't update my PR.  Do I need to trigger something manually?

14rcole commented:
Never mind, it seems to have just been delayed

nalind commented on cmd/buildah/common.go:
```diff
@@ -1,15 +1,28 @@
+	Blobs          []types.BlobInfo    `json:"blob-list"`
+	Layers         map[string][]string `json:"layers"`
+	SignatureSizes []string            `json:"signature-sizes"`
+}
```
I wouldn't recommend parsing the metadata that the image library records - it intentionally doesn't make that a public type, so that it can change the layout to fit its needs.  In the future, it very likely will be missing the information we're trying to read, since it will be duplicating information that newer versions of the storage library will cache in publicly-defined structures.

nalind replied:
We can _probably_ get away with this for now, but when we pull in a newer storage library, we'll want to remember to switch to using the public information.

14rcole replied:
What would you suggest as an alternative?  Something like unmarshaling the data to a `map[string]interface{}`?

nalind replied:
We're not providing anything better (yet), so this is probably going to have to do until we update the storage and image libraries to provide this information as part of their public APIs.

nalind commented on cmd/buildah/images.go:
```diff
@@ -13,6 +18,23 @@ type jsonImage struct {
+	dangling         string
+	label            string
+	beforeImage      string // Images are sorted by date, so we can just output until we see the image
+	sinceImage       string // Images are sorted by date, so we can just output until we don't see the image
```
The storage library doesn't take steps to ensure anything about the ordering of images that it returns, and it's probably a bug that the creation date that the image library metadata records for a pulled image is the time at which the image was pulled -- if the image's config blob includes a datestamp indicating when it was built on whichever system it was built, I think we'd want to use that value instead.

14rcole replied:
Okay, I'll fix this

nalind commented on cmd/buildah/images.go:
```diff
@@ -24,22 +46,35 @@ var (
 	}
-	imagesDescription = "Lists locally stored images."
+
+	imagesDescription = "lists locally stored images."
```
This makes the capitalization stop matching the other commands.  What's the intent here?

14rcole replied:
These were all unintentional changes, I'll revert them

nalind commented on cmd/buildah/images.go:
```diff
@@ -24,22 +46,35 @@ var (
 	imagesCommand     = cli.Command{
 		Name:        "images",
-		Usage:       "List images in local storage",
+		Usage:       "list images in local storage",
```
This makes the capitalization stop matching the rest of the `buildah --help` output.  What's the intent here?

nalind commented on cmd/buildah/images.go:
```diff
@@ -24,22 +46,35 @@ var (
 		Flags:       imagesFlags,
 		Action:      imagesCmd,
-		ArgsUsage:   " ",
+		ArgsUsage:   "",
```
This causes the help output to include `[arguments...]` after `[command options]` in the usage section.  It's preferable to indicate what type of value we expect to be given, as we do for the other commands.

nalind commented on cmd/buildah/images.go:
```diff
@@ -24,22 +46,35 @@ var (
 		},
 		cli.BoolFlag{
-			Name:  "notruncate",
+			Name:  "no-trunc, notruncate",
```
What's the rationale for adding this alias for this option to this command, but not to any of the other commands that offer a `--notruncate` option?

14rcole replied:
I hadn't looked at any of the other commands.  When I wrote this code for kpod, it was supposed to match the docker cli, which uses `--no-trunc`.  I chose not to remove the option when I ported the code to buildah.  I can remove it or add the option to the other commands if you would like

nalind replied:
@rhatdan You've got stronger opinions on the CLI than I do.  WDYT?

rhatdan replied:
I think we should match the CLI of kpod and docker.   No value in the alternate spelling.

14rcole replied:
Should I change this in the other commands that offer a `--notruncate` option then?  Or should that be a separate PR?

rhatdan replied:
Lets open separate pull requests.

14rcole commented:
@rhatdan is there a way to manually re-trigger the checks?  It looks like an internal error occurred on the Red Hat CI

rhatdan commented:
@14rcole I usually just do a 
```
git commit -a --amend -s
git push rhatdan --force
```

rhatdan commented:
@nalind you ok with merging?

rhatdan commented:
@14rcole you need to update buildah.md and README.md with these new commands.

14rcole commented:
@rhatdan The commands were already listed in buildah.md and README.md, and I updated both buildah-images.md and buildah-rmi.md to include the flags I added.  Is there something else in those files that needs to be changed?

rhatdan commented:
Ok Sorry I thought they were missing.

rhatdan commented:
I am going to merge...
@rh-atomic-bot r+ c7c207b

TomSweeneyRedHat commented:
Also fixes #232
